src/io/flutter/editor/FlutterSaveActionsManager.java

```diff
@@ -0,0 +1,195 @@
+import com.intellij.util.PsiErrorElementUtil;
+import com.intellij.util.messages.MessageBus;
+import com.intellij.util.messages.MessageBusConnection;
+import com.jetbrains.lang.dart.DartLanguage;
+import com.jetbrains.lang.dart.analyzer.DartAnalysisServerService;
+import com.jetbrains.lang.dart.assists.AssistUtils;
+import io.flutter.FlutterUtils;
+import io.flutter.dart.DartPlugin;
+import io.flutter.settings.FlutterSettings;
+import org.dartlang.analysis.server.protocol.SourceEdit;
+import org.dartlang.analysis.server.protocol.SourceFileEdit;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+import java.util.List;
+
+/**
+ * A manager class to run actions on save (formatting, organize imports, ...).
+ */
+public class FlutterSaveActionsManager {
+  private static final Logger LOG = Logger.getInstance(FlutterSaveActionsManager.class);
+
+  /**
+   * Initialize the save actions manager for the given project.
+   */
+  public static void init(@NotNull Project project) {
+    // Call getInstance() will init FlutterFormatManager for the given project.
+    getInstance(project);
+  }
+
+  public static FlutterSaveActionsManager getInstance(@NotNull Project project) {
+    return ServiceManager.getService(project, FlutterSaveActionsManager.class);
+  }
+
+  private final @NotNull Project myProject;
+
+  private FlutterSaveActionsManager(@NotNull Project project) {
+    this.myProject = project;
+
+    final MessageBus bus = project.getMessageBus();
+    final MessageBusConnection connection = bus.connect();
+    connection.subscribe(AppTopics.FILE_DOCUMENT_SYNC, new FileDocumentManagerAdapter() {
+      @Override
+      public void beforeDocumentSaving(@NotNull Document document) {
+        handleBeforeDocumentSaving(document);
+      }
+    });
+  }
+
+  private void handleBeforeDocumentSaving(@NotNull Document document) {
+    final FlutterSettings settings = FlutterSettings.getInstance();
+    if (!settings.isFormatCodeOnSave()) {
+      return;
+    }
+
+    if (!myProject.isInitialized() || myProject.isDisposed()) {
+      return;
+    }
+
+    final VirtualFile file = FileDocumentManager.getInstance().getFile(document);
+    if (file == null) {
+      return;
+    }
+
+    if (!FlutterUtils.isDartFile(file)) {
+      return;
+    }
+
+    final PsiFile psiFile = PsiDocumentManager.getInstance(myProject).getPsiFile(document);
+    if (psiFile == null || !psiFile.isValid()) {
+      return;
+    }
+
+    final Module module = getModuleForPath(file);
+    if (module == null) {
+      return;
+    }
+
+    if (!DartPlugin.isDartSdkEnabled(module)) {
+      return;
+    }
+
+    // check for errors
+    if (PsiErrorElementUtil.hasErrors(myProject, psiFile.getVirtualFile())) {
+      return;
+    }
+
+    DartAnalysisServerService.getInstance(myProject).serverReadyForRequest(myProject);
+
+    if (settings.isOrganizeImportsOnSaveKey()) {
+      performOrganizeImportsOnSave(document, file);
+    }
+    else {
```

Comment:
So, we do either organize import, or format?
Maybe I don't understand IntelliJ UI format, but it says `JCheckBox`, so both can be checked?

Reply:
Ah, I see.
We do both when organize imports.

Reply:
I renamed the methods to make this clearer.